### ISSUE TITLE
urlpath doesn't do any quoting

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2625]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#2625|
|Type|defect|
|Created|2007-05-04 15:38:34Z|
|Branch|https://github.com/twisted/twisted/tree/urlpath-quote-2625|

e.g. 

```
>>> str(urlpath.URLPath().child("hello/ there"))
'http://localhost/hello/ there'
```

ought to be

```
>>> str(urlpath.URLPath().child("hello/ there"))
'http://localhost/hello%2f%20there'
```

Attachments:

* [2625.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ccb/ccbe0111e5eb9cc42540d4f93e0e8fed61b230e0/650299f52b724392d3c8ef13b48815b28bf5a1f0.patch) (1402 bytes) - added by truekonrads on 2009-08-27 19:07:52Z - 
* [2625.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ccb/ccbe0111e5eb9cc42540d4f93e0e8fed61b230e0/533401b567db16bb4d72b987f85a24dbf824af0d.patch) (2642 bytes) - added by truekonrads on 2009-08-27 20:00:11Z - Improved version that addresses exarkun's comments
* [2625.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ccb/ccbe0111e5eb9cc42540d4f93e0e8fed61b230e0/ae071cb16352bb041f1680b8d354655cd83094a2.patch) (8637 bytes) - added by danben on 2010-05-12 21:58:22Z - 
* [2625.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ccb/ccbe0111e5eb9cc42540d4f93e0e8fed61b230e0/3771e0be2d547e7aab52a94d49eaea29c110e4e2.patch) (9432 bytes) - added by danben on 2010-05-12 22:12:43Z - 
* [urlpath-quote.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ccb/ccbe0111e5eb9cc42540d4f93e0e8fed61b230e0/4aaeb7ddae6bcd29c1578ae990a6cc19fff6c276.patch) (1483 bytes) - added by taos on 2011-01-26 13:22:08Z - Added some new docstrings, removed __str__ method in favor of toString(), not done unit tests yet.
* [2625.6.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ccb/ccbe0111e5eb9cc42540d4f93e0e8fed61b230e0/9d4b3d43b5252d0bc9d33d145bee6206fd0c28d8.patch) (7029 bytes) - added by magmatt on 2013-08-08 15:43:56Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2625 2625
type__defect defect
reporter__itamarst itamarst
priority__normal normal
milestone__ 
branch__branches_urlpath_quote_2625 branches/urlpath-quote-2625
branch_author__truekonrads truekonrads
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1178293114000000 1178293114000000
changetime__1481462099154513 1481462099154513
version__None None
owner__magmatt magmatt
cc__truekonrads cc__exarkun cc__danben cc__mithrandi cc__magmatt
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @itamarst|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|truekonrads commented|
|-|-|

I implemented the child() thing. Maybe we need to handle more stuff, like url in __init__?
I am putting this up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to truekonrads|
|-|-|

Thanks for the patch!  Here are a few comments:

  1. `URLPath.child` should have a docstring.  Can you add one?  One thing it should mention, in particular, is that the path argument should not be quoted.
  1. Operators should have one space around them.  For example, `fixedPath=foo` should be `fixedPath = foo`.  Similarly, function arguments should be separated from each other by a ", ", not just a ",".
  1. You can make `urllib.quote` do _all_ of the work for you by passing an empty string as the second argument.  This way you won't have to use `str.replace` afterwards.
  1. `test_childQuotes` should also have a docstring.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|truekonrads commented|
|-|-|

Replying to [exarkun](#note_5):

Done.

Additionally, I've replaced the default value of keepQuery=0 to keepQuery=False as it is a boolean condition anyway.

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|truekonrads removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [27252]) Branching to 'urlpath-quote-2625'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [27253]) Apply 2625.2.patch

refs [#2625](https://github.com/twisted/twisted/issues/2625)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [27254]) Clean up some docstring language; fix some minor coding standard violations

refs [#2625](https://github.com/twisted/twisted/issues/2625)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

I made some minor changes to the new patch; mostly whitespace and docstring adjustments.  Checking out [how it does on buildbot] now.  Assuming it goes well, I'll merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to truekonrads<br>@exarkun set status to `new`|
|-|-|

Ahem, [how it does on buildbot](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/urlpath-quote-2625)

However, meh.  I changed my mind.  This is an incompatible change which will break any existing careful uses of `URLPath.child`.  The existing method and behavior should be left alone, but deprecated.  A new method should be introduced which has the desired semantics (which are now well implemented, I think, in the branch by the `child` method).  Additionally, `URLPath.sibling` should probably receive similar treatment, and the rest of the methods checked for this problem.

In general, please refer to CompatibilityPolicy.  This document isn't complete or finalized, but it's a good guideline nonetheless, and we should adhere to it as closely as possible.

**Comment by twisted-trac** [comment]
|<img alt="sah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|sah set owner to sah<br>sah set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="sah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|sah set owner to truekonrads<br>sah set status to `new`|
|-|-|

Yikes, nevermind, reading comments I realized this involves too much policy for me right now.

**Comment by twisted-trac** [comment]
|<img alt="danben's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|danben removed owner|
|-|-|

Building on the patches submitted by truekonrads, I made an attempt to address exarkun's compatibility concerns.  Also included are some minor changes like test case names and boolean default arguments instead of ints.

In addition to ```URLPath.child```, I provided quoting implementations of ```URLPath.sibling``` and ```URLPath.fromString```.  The names are the best I could come up with but I feel there's definite room for improvement there.

In unit testing the deprecated methods, I'm not sure if what I have is sufficient (counting the number of warnings seen) or if more detail is needed.

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|truekonrads commented|
|-|-|

I think that URLPath class with correct quoting should be called differently from old URLPath not added new methods. This comes to mind:

 * SafeURLPath
 * URLPath2
 * URLTrek
 * NewURLPath (which will be followed by ReallyNewThisTimeURLPath :) )

Usage of old URLPath should be phased out from Twisted and warn raised from __init__.
P.S. I think this is funny how long it takes to close an easy ticket.

**Comment by twisted-trac** [comment]
|<img alt="danben's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|danben commented|
|-|-|

Do you think this should be a new class in the same module or something like ```urlpath2.URLPath```?

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|truekonrads commented|
|-|-|

Replying to [danben](#note_17):
> Do you think this should be a new class in the same module or something like ```urlpath2.URLPath```?

Yes, I think this is a nice solution - urlpath2.URLPath. I would suggest you also add DepreciationWarning to the urlpath module.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

Hang on a minute.  Why is a new module a better idea than introducing new methods on the existing class?

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|truekonrads commented|
|-|-|

Replying to [exarkun](#note_19):
> Hang on a minute.  Why is a new module a better idea than introducing new methods on the existing class?
Because it is less confusing to the programmer - explorer who discovers a module, looks up its methods and applies them. He does not have to guess if he/she should use child() or safeChild() or newChild() or child(). urlpath2 vs urlpath immediately suggests to use the newer version.

**Comment by twisted-trac** [comment]
|<img alt="danben's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|danben commented|
|-|-|

Replying to [truekonrads](#note_20):
> Replying to [exarkun](#note_19):
> > Hang on a minute.  Why is a new module a better idea than introducing new methods on the existing class?
> Because it is less confusing to the programmer - explorer who discovers a module, looks up its methods and applies them. He does not have to guess if he/she should use child() or safeChild() or newChild() or child(). urlpath2 vs urlpath immediately suggests to use the newer version.

I agree with this, assuming that we want quoting to be the default behavior.  One drawback that I can see is a surprise for users who upgrade from ```urlpath``` to ```urlpath2``` without looking carefully at the documentation to see that the methods behave differently, but maybe this could be made explicit at the class-level documentation since ```URLPath``` is not a large class.

Anyway, I will defer to experience here.  exarkun, let me know which approach you prefer and I'll make a patch based on that.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

So, -1 on a new module.  _especially_ not "urlpath2".

I do think that making this a new class rather than a new collection of methods would be a good idea.  I say this because it affects the behavior of the `fromString`, `child`, and `sibling` methods, and having a suffixed version of each method is somewhat verbose.

I suggest adding a new _class_, `urlpath.URL`, which has the right quoting behavior on all of its methods, and deprecating `URLPath`.

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_23"></a>|truekonrads commented|
|-|-|

Sounds like a solution. urlpath.URL + depreciationwarning has my vote.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun set owner to truekonrads|
|-|-|

So, to summarize:

  * We don't want to fix `URLPath`
  * We want to introduce a new API which doesn't have the bugs `URLPath` has
    * Quoting should be handled in a way that promotes secure usage
    * [ticket:2093 All 4 parts of the netloc should be parsed]
    * [ticket:2094 Query arguments should be easily inspected and manipulated]
    * [http://divmod.org/trac/ticket/2409 unicode should be supported]
    * [http://divmod.org/trac/ticket/1389 Both absolute and relative URLs should be supportable]

I suggest that this start with _just_ a model which can represent all of the information present in a URL (or IRI).  Adding useful methods for manipulating the state can come afterwards.  And eventually we'll need to have an API on `Request` for getting one of these _new_ things, rather than an old `URLPath`.  Then we can deprecate `URLPath`.

If people actually agree this is the way to go, then this ticket should be closed as wontfix and at least one new one opened describing the new thing that's going to be implemented.

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_26"></a>|truekonrads set status to `closed`|
|-|-|

Replying to [exarkun](#note_25):
> If people actually agree this is the way to go, then this ticket should be closed as wontfix and at least one new one opened describing the new thing that's going to be implemented.

Sounds like a decent plan. Will you do the honors then?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph set status to `reopened`|
|-|-|

Let's change the summary, not close as wontfix.  The problem still exists: Twisted doesn't have a good abstraction for generating quoted URLs.  Also, new tickets should be opened before old ones are closed, so we won't forget to do it.

In any event, I have another suggestion besides making a new class: `__str__` is a bad way to get a canonicalized string format representation, we should really deprecate _that_.  Why not just do the quoting at the point of stringification, rather than in the `child()` methods, and add a new, good method for stringification?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_29"></a>|@itamarst commented|
|-|-|

Unfortunately adding an alternative to `__str__` isn't good enough, since e.g. `sibling()` takes what most users would expect to be an unquoted path, and then attaches it to quoted path elements.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun commented|
|-|-|

> Let's change the summary, not close as wontfix.

I notice the summary was never changed. :)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_31"></a>|@itamarst commented|
|-|-|

I tried making new implementation. It's hard. Easier short term solution may be to just have two modes in URLPath, backwards compatible but deprecated mode, and correct quoting mode.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

What's hard?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_33"></a>|@itamarst commented|
|-|-|

Unicode ([#5388](https://github.com/twisted/twisted/issues/5388)). Bytes-only isn't hard (and implementation-wise may well be quite close to having two different modes on same class).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_34"></a>|@glyph commented|
|-|-|

Replying to [itamar](#note_33):
> Unicode ([#5388](https://github.com/twisted/twisted/issues/5388)). Bytes-only isn't hard (and implementation-wise may well be quite close to having two different modes on same class).

The API should just expose a new name so it's easy to know what "mode" - by which I mean "behavior" - your client code expects.  The guts of the implementation isn't super interesting.

**Comment by twisted-trac** [comment]
|<img alt="magmatt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_35"></a>|magmatt commented|
|-|-|

After chatting with itamar about this, we want to leave the unicode problem for [#5388](https://github.com/twisted/twisted/issues/5388).  So this ticket is *only* to properly quote things like '/' and ' ' (and other ascii characters that need quoting).

The implementation will involve accepting a flag passed to URLPath.__init__ that will turn on quoting.  Also, there will be a subclass of URLPath that has the flag turned on by default.

**Comment by twisted-trac** [comment]
|<img alt="magmatt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_36"></a>|magmatt commented|
|-|-|

The patch I just added 2625.6.patch uses code from the urlpath-quote-2625 branch, but is based on trunk (instead of 3-year-old trunk)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst removed owner<br>@itamarst set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_38"></a>|@jonathanj set owner to magmatt|
|-|-|

Thanks for updating the patch, magmatt!

  1. The first-person narrative docstring style is no longer preferred, please rephrase docstrings in this style to just talk about their behaviour without the narrative or first-person-ness.

  2. Would you add a nice helpful docstrings for `URLPath.sibling`, `URLPath.parent` and `URLPath.here` too, since their behaviour is changing in this patch too?

  3. It would be preferable if the docstrings in the tests could use the linking syntax, instead of the code literal syntax, when talking about methods. ie. `C{child}` becomes `L{URLPath.child}`.

  4. Please use `assertEqual` instead of `assertEquals`.

  5. (This is mostly just commentary.) It's unfortunate that passing an unquoted URL to `URLPath.fromString` is broken, regardless of whether you pass `quote=True`; I realise it is at least documented this way but it sure looks weird (maybe `fromString` is a misleading name compared to something like `fromURIBytes` or `fromUnquotedString`):

```
>>> str(URLPath.fromString('http://example.com/foo bar', quote=True))
'http://example.com/foo bar'
```

  6. Since there is now a docstring for `URLPath.__init__`, could you document the other parameters too?

  7. I don't think there is a need to redundantly document the default value of `quote` in `URLPath.__init__`.

  8. The docstring for `URLPath` itself should at least include a one-liner description of the class.

  9. Some docstrings and parameters are missing periods at the end of their sentences.

  10. Would you make sure there are two blank lines between methods where the code (or surrounding code) is being changed?

Sorry about all the nitpicking, I hope it doesn't deter you from getting this patch landed.

**Comment by twisted-trac** [comment]
|<img alt="magmatt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_39"></a>|magmatt commented|
|-|-|

Replying to [jonathanj](#note_38):
>   1. The first-person narrative docstring style is no longer preferred, please rephrase docstrings in this style to just talk about their behaviour without the narrative or first-person-ness.
Have we always been at war with Eastasia?  What is the preferred style (is there a wiki page or something I can read)?
 
> Sorry about all the nitpicking, I hope it doesn't deter you from getting this patch landed.
Honestly, it does deter me.  I feel like I'm playing Guess What the Reviewer Wants.  I wish reviewers could easily make little fixes (and have them reviewed).  Some reviews would be nicer as a patch than as English.  But I suppose such a system would have downsides too.

I'll see if I can get to this (but it won't be today).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_40"></a>|@glyph commented|
|-|-|

Since the issue of frustration came up, I figured I'd do a meta-review! :-)

My main point is: please separate suggestions and requirements into distinct sections so that the author has a clear idea of what they actually _need_ to do to get it landed, versus the things they might want to consider, either for future tickets or at their own discretion before re-review.

Replying to [jonathanj](#note_38):
> Thanks for updating the patch, magmatt!
> 
>   1. The first-person narrative docstring style is no longer preferred, please rephrase docstrings in this style to just talk about their behaviour without the narrative or first-person-ness.

This is useful as a suggestion, but despite the fact that this style isn't preferred, having documentation in this style is still better than not having documentation at all.  Not mandatory.

>   2. Would you add a nice helpful docstrings for `URLPath.sibling`, `URLPath.parent` and `URLPath.here` too, since their behaviour is changing in this patch too?

Changed behavior does require changed documentation (and changed tests) so this would be a requirement.

>   3. It would be preferable if the docstrings in the tests could use the linking syntax, instead of the code literal syntax, when talking about methods. ie. `C{child}` becomes `L{URLPath.child}`.

Suggestion, but hopefully trivial enough that magmatt would want to do it :).  (Also, for your information, `L{child <URLPath.child>}` will make the link appear as just the word "child" but still link to the right method.

>   4. Please use `assertEqual` instead of `assertEquals`.

This should really come along with a citation.  How did you know that this was correct?

>   5. (This is mostly just commentary.) It's unfortunate that passing an unquoted URL to `URLPath.fromString` is broken, regardless of whether you pass `quote=True`; I realise it is at least documented this way but it sure looks weird (maybe `fromString` is a misleading name compared to something like `fromURIBytes` or `fromUnquotedString`):

"Mostly just commentary" is not an actionable statement :-).  Should be in a separate "commentary" section.

>   6. Since there is now a docstring for `URLPath.__init__`, could you document the other parameters too?

Doable as a separate ticket, since the parameters aren't new, right?

>   7. I don't think there is a need to redundantly document the default value of `quote` in `URLPath.__init__`.

Definitely not a regression, so not necessary to deal with.  Honestly, redundant documentation isn't such a bad thing.  It's more work for the author but nice for the reader, since it gives them multiple angles to approach the documentation from.

>   8. The docstring for `URLPath` itself should at least include a one-liner description of the class.

I am not so sure about this one.

>   9. Some docstrings and parameters are missing periods at the end of their sentences.

Again, this is acceptable.

>   10. Would you make sure there are two blank lines between methods where the code (or surrounding code) is being changed?

Despite being nitpicky, coding standard compliance is required (although if that's the _only_ issue, the reviewer can make this change themselves).

(But this should also come with a reference citation...)

> Sorry about all the nitpicking, I hope it doesn't deter you from getting this patch landed.

**Comment by twisted-trac** [comment]
|<img alt="magmatt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_41"></a>|magmatt commented|
|-|-|

In IRC exarkun pointed out that danben's patch (2625.4.patch) is probably a better direction (adding new methods that do quoting instead of adding a _quote attribute and attempting to change existing functions).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_42"></a>|@glyph set status to `closed`|
|-|-|

This was fixed as a side-effect of [#5388](https://github.com/twisted/twisted/issues/5388), in f8b827389fc346535afba5794fc72976f1235996.
